pytest_playwright/pytest_playwright.py

```diff
@@ -145,22 +181,89 @@ def browser(launch_browser: Callable[[], Browser]) -> Generator[Browser, None, N
     browser = launch_browser()
     yield browser
     browser.close()
+    artifacts_folder.cleanup()
 
 
 @pytest.fixture
 def context(
-    browser: Browser, browser_context_args: Dict
+    browser: Browser,
+    browser_context_args: Dict,
+    pytestconfig: Any,
+    request: pytest.FixtureRequest,
 ) -> Generator[BrowserContext, None, None]:
+    pages: List[Page] = []
     context = browser.new_context(**browser_context_args)
+    context.on("page", lambda page: pages.append(page))
+
+    tracing_option = pytestconfig.getoption("--tracing")
+    capture_trace = tracing_option in ["on", "retain-on-failure"]
+    if capture_trace:
+        context.tracing.start(
+            name=slugify(request.node.nodeid),
+            screenshots=True,
+            snapshots=True,
+        )
+
     yield context
+
+    if capture_trace:
+        retain_trace = tracing_option == "on" or (
+            request.node.rep_call.failed and tracing_option == "retain-on-failure"
+        )
+        if retain_trace:
+            trace_path = _build_artifact_test_folder(pytestconfig, request, "trace.zip")
+            context.tracing.stop(path=trace_path)
+        else:
+            context.tracing.stop()
+
+    screenshot_option = pytestconfig.getoption("--screenshot")
+    capture_screenshot = screenshot_option == "on" or (
+        request.node.rep_call.failed and screenshot_option == "only-on-failure"
+    )
+    if capture_screenshot:
+        for index, page in enumerate(pages):
+            human_readable_status = (
+                "failed" if request.node.rep_call.failed else "finished"
+            )
+            screenshot_path = _build_artifact_test_folder(
+                pytestconfig, request, f"test-{human_readable_status}-{index+1}.png"
+            )
+            try:
```

Comment:
Nit: A more pythonic one is https://docs.python.org/3/library/contextlib.html#contextlib.suppress